### PR TITLE
[state_sync] Part 2: Cleanup `process_snapshot` impl in chain

### PR DESCRIPTION
**Pull request by shreyan-gupta**
This PR cleans up the implementation of calling make_snapshot_callback in chain. 

No functional changes

### CODE REVIEW

File: chain/chain/src/block_processing_utils.rs [hunk 1]
```diff
@@ -29,8 +29,6 @@ pub(crate) struct BlockPreprocessInfo {
     pub(crate) apply_chunks_done: Arc<OnceCell<()>>,
     /// This is used to calculate block processing time metric
     pub(crate) block_start_processing_time: Instant,
-    /// Whether needs to create a state snapshot after processing this block.
-    pub(crate) need_state_snapshot: bool,
```
Comment (shreyan-gupta) [inline chain/chain/src/block_processing_utils.rs:33]: We don't need snapshot details in BlockPreprocessInfo as it's only used in the preprocess function later (and not after post processing)

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -751,10 +745,10 @@ impl Chain {
             last_time_head_updated: StaticClock::instant(),
             pending_state_patch: Default::default(),
             requested_state_parts: StateRequestTracker::new(),
-            state_snapshot_helper: make_snapshot_callback.map(|callback| StateSnapshotHelper {
-                make_snapshot_callback: callback,
-                test_snapshot_countdown_and_frequency,
-            }),
+            state_snapshot_helper: StateSnapshotHelper {
+                make_snapshot_callback,
+                blocks_since_last_snapshot: 0,
```
Comment (shreyan-gupta) [inline chain/chain/src/chain.rs:750]: Changed from countdown to count up mechanism

**Review by nikurt — APPROVED** [review]
:+1: :+1: :+1:

**Comment by wacban** [comment]
@shreyan-gupta jfyi some checks in CI are failing

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -4279,42 +4276,58 @@ impl Chain {
...
+            // For every epoch, we snapshot if the next block would be in a different epoch
+            StateSnapshotType::EveryEpoch => is_epoch_boundary,
+            // For resharding only, we snapshot if next block would be in a different shard layout
+            StateSnapshotType::ForReshardingOnly => is_epoch_boundary && will_shard_layout_change,
+            // For every N blocks, we snapshot if next block would be in a different epoch
+            // OR if we have reached N blocks since the last snapshot
+            StateSnapshotType::EveryEpochAndNBlocks(n) => {
+                is_epoch_boundary || self.state_snapshot_helper.blocks_since_last_snapshot >= n
             }
+        };
```
Comment (wacban) [inline chain/chain/src/chain.rs:4324]: <3 
very nice and neat

File: core/store/src/trie/shard_tries.rs [hunk 4]
```diff
@@ -216,7 +216,6 @@ impl ShardTries {
         self.0.flat_storage_manager.clone()
     }
 
-    // TODO (#9989): Change visibility to crate once we remove state_snapshot_every_n_blocks from chain
```
Comment (wacban) [inline core/store/src/trie/shard_tries.rs:219]: Is it no longer relevant?

**Review by wacban — COMMENTED** [review]
LGTM, please ping me when ready to be merged and I'll approve

File: chain/chain/src/chain.rs (hunk 3 above)
Comment (shreyan-gupta) [reply chain/chain/src/chain.rs:4317]: Thanks!

File: core/store/src/trie/shard_tries.rs (hunk 4 above)
Comment (shreyan-gupta) [reply core/store/src/trie/shard_tries.rs:219]: Yes, I realized we can't remove the functionality of "when to snapshot" from chain as we need epoch manager; which means we would need to expose the snapshot config from trie.

**Pull request merged**